### PR TITLE
[ML][Transforms] adjusting when and what to audit

### LINKED ISSUE #45834: [ML] Transform audit message improvements
Feedback being given on 8.0.0 `{
  "build" : {
    "hash" : "bbea61a232db278db2941e726bbfec16bea8d7f6",
    "date" : "2019-08-21T14:20:38.772853Z"
  },`

// missing Insufficient memory message

Continuous transform audit messages no longer report the following message. This was a useful message in that it indicates a corrective action being made on the part of the transform and if being done frequently, will indicate resources are constrained. The following is still logged, but not audited. Where did it go? Can we have it back?

`Data frame transform [transform-01]:Insufficient memory for search, reducing number of buckets per search from [500] to [365]`

// frequency of messages

In a system with long running continuous data frames and index creation / deletion, there is a periodic `Failed to retrieve checkpoint`.

image

Looking in the logs, this is caused by failure to retrieve checkpoint info caused by a global mismatch. We retry.

1. Do we need to show this to the user? If the retries fail then we should. I'm not sure this is worth auditing. If we think it is, then would be good for the message that is presented to the user to indicate that a retry occurs.

2. We only show this audit message the first time it occurs. Subsequent similar messages throughout the night were not audited (but they were in the logs). What is the logic for the frequency of auditing this? (I saw 9 log entries for one transform but only 1 audit).

// Failed transform

In the case below, the number of buckets per search was set to 10. A CBE occurred due to load in the cluster from other processes. The 10 subsequent retries were fired in quick succession and also caused CBEs [1]. The final audit message indicated that the cluster state could not be updated. The "useful" audit message would have been that there was "insufficient memory unable to continue" which was the log message before. 

(note that this may have been due to the unfortunate coincidence of two failures - both could be audited, with the insufficient memory message being more informative).

image

```
[2019-08-22T05:17:53,886][ERROR][o.e.x.d.t.DataFrameTransformTask] [node2] Data frame transform [transform-05]: Insufficient memory for search after repeated page size reductions to [10], unable to continue pivot, please simplify job or increase heap size on data nodes.
[2019-08-22T05:17:54,066][ERROR][o.e.x.d.t.DataFrameTransformTask] [node2] Failed to update state for data frame transform [transform-05]
org.elasticsearch.transport.RemoteTransportException: [node1][127.0.0.1:9351][cluster:admin/persistent/update_status]
```
cc @benwtrent 

[1] A topic for a different discussion on throttling and back off.

### PR DESCRIPTION
**Pull request by benwtrent**
This PR:
* Adds failure reason for checkpoint creation|retrieval failure
* Audits that we will retry certain failures
* Logs the state and reason when it fails to write to cluster state.

closes #45834

### CODE REVIEW

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java [hunk 1]
```diff
@@ -1101,6 +1103,15 @@ synchronized void handleFailure(Exception e) {
                     "task encountered irrecoverable failure: " + e.getMessage() :
                     "task encountered more than " + transformTask.getNumFailureRetries() + " failures; latest failure: " + e.getMessage();
                 failIndexer(failureMessage);
+            } else {
+                // Since our schedule fires again very quickly after failures it is possible to run into the same failure numerous
+                // times in a row, very quickly. We do not want to spam the audit log with repeated failures, so only record the first one
+                if (e.getMessage().equals(lastAuditedExceptionMessage) == false) {
+                    auditor.warning(transformTask.getTransformId(),
+                        "Data frame transform encountered an exception: " + e.getMessage() +
+                            " Will attempt again at next scheduled trigger.");
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java]: nit: Do we need a dot after the message? It seems like 2 different sentences.

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM Left a suggestion/question for phrasing one of the messages but it's ready to go regardless.

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java (hunk 1 above)
Comment (benwtrent) [reply x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java]: @dimitris-athanasiou I waffled on that. But certain Messages already had periods, then it turned into `Data frame ... exception: Something went wrong.. Will attempt... trigger.` which looked more strange to me than just not having a period at all.

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java [hunk 2]
```diff
@@ -459,7 +460,10 @@ synchronized void markAsFailed(String reason, ActionListener<Void> listener) {
         persistStateToClusterState(newState, ActionListener.wrap(
             r -> listener.onResponse(null),
             e -> {
-                logger.error(new ParameterizedMessage("[{}] failed to set task state as failed to cluster state.", getTransformId()),
+                String msg = "Failed to persist to cluster state while marking task as failed with reason [" + reason + "].";
```
Comment (davidkyle) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java]: Unrelated: In this function `markAsFailed we audit a message if the DF is failed while it is stopping (L432). That is definitely a log-able event as it aids understanding but I'm not sure it should be audited.

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java [hunk 3]
```diff
@@ -1050,13 +1048,17 @@ protected void createCheckpoint(ActionListener<DataFrameTransformCheckpoint> lis
...
+                                    new RuntimeException("Failed to create checkpoint due to " + createCheckpointException.getMessage(),
+                                        createCheckpointException));
                             }
                     )),
                     getCheckPointException -> {
                         logger.warn(new ParameterizedMessage("[{}] failed to retrieve checkpoint.", transformId),
                             getCheckPointException);
-                        listener.onFailure(new RuntimeException("Failed to retrieve checkpoint", getCheckPointException));
+                        listener.onFailure(
+                            new RuntimeException("Failed to retrieve checkpoint due to " + getCheckPointException.getMessage(),
```
Comment (davidkyle) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java]: Won't the exception message be repeated here if the exception also the cause?

Comment (benwtrent) [reply x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java]: @davidkyle the internal cause for sure, but in the handling paths we only make `Exception#getMessage()` visible to the user through an audit message.

**Review by davidkyle — APPROVED** [review]
LGTM

**Pull request merged**